modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsManagerImpl.java

```diff
@@ -278,4 +614,373 @@ private ObjectStatisticsImpl aggregateLocalStatistics(
+            }
+            catch (IgniteCheckedException e) {
+                log.warning(String.format("Unable to complete request %s due to error %s", req.reqId(), e.getMessage()));
+                // TODO: send cancel to originator node
+            }
+        }
+
+        StatsCollectionResponse res = new StatsCollectionResponse(req.colId(), req.reqId(), collected);
+
+        if (locNode.equals(nodeId))
+            receiveLocalStatistics(nodeId, res);
+        else {
+            try {
+                ctx.io().sendToCustomTopic(nodeId, TOPIC, res, GridIoPolicy.QUERY_POOL);
+            }
+            catch (IgniteCheckedException e) {
+                if (log.isDebugEnabled())
+                    log.debug(String.format(
+                            "Unable to send statistics collection result to node %s in response to colId %s, reqId %s",
+                        nodeId, req.colId(), req.reqId()));
+            }
+
+            // Not local collection - remove by its reqId.
+            currCollections.updateCollection(req.reqId(), s -> null);
+        }
+    }
+
+    /**
+     * Schedule statistics collection by specified request.
+     *
+     * @param nodeId request origin node.
+     * @param msg request message.
+     */
+    private void handleCollectionRequest(UUID nodeId, StatsCollectionRequest msg) {
+        assert msg.reqId() != null : "Got statistics collection request without request id";
+
+        currCollections.addActiveCollectionStatus(msg.reqId(), new StatCollectionStatus(msg.colId(), msg.keys().keySet(), null));
+
+        statMgmtPool.submit(() -> processLocal(nodeId, msg));
+    }
+
+    /**
+     * Send requests to target nodes except of local one.
+     *
+     * @param reqs Collection of addressed requests to send.
+     * @return Collection of addressed requests that has errors while sending or {@code null} if all requests was send
+     * successfully.
+     */
+    private <T extends Message> Collection<StatsAddrRequest<T>> sendRequests(Collection<StatsAddrRequest<T>> reqs) {
+        UUID locNode = ctx.localNodeId();
+        Collection<StatsAddrRequest<T>> res = null;
+
+        for (StatsAddrRequest<T> req : reqs) {
+            if (locNode.equals(req.nodeId()))
+                continue;
+
+            try {
+                ctx.io().sendToCustomTopic(req.nodeId(), TOPIC, req.req(), GridIoPolicy.QUERY_POOL);
+            }
+            catch (IgniteCheckedException e) {
+                if (res == null)
+                    res = new ArrayList<>();
+
+                res.add(req);
+            }
+        }
+
+        return res;
+    }
+
+    /**
+     * Handle node left event:
+     * 1) Cancel all collection tasks which expect specified node statistics result.
+     * 2) Cancel collection task requested by node left.
+     *
+     * @param nodeId leaved node id.
+     */
+    private void onNodeLeft(UUID nodeId) {
+        Map<UUID, Map<StatsKeyMessage, int[]>> failedCollections = new HashMap<>();
```

Comment:
failedCollections is not used

Reply:
fixed in StatisticsGatheringRequestCrawlerImpl